### PR TITLE
Add VK_KHR_synchronization2 example

### PR DESCRIPTION
**Pull request by SaschaWillems**
## Description

This PR adds a "new" example that shows usage of the new synchronization features introduced with `VK_KHR_synchronization2`. The sample is based on the compute shader N-Body sample, and allows an easy comparison between the old and new synchronization api.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [ ] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

## Notes
- I haven't tested with Validation layers yet as the accompanying PR has not yet been merged (https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/2554)
- Tested on an NVIDIA RTX 2060 on Windows 10
- This PR also updates Volk and the Vulkan headers to 1.2.170

### CODE REVIEW

**Comment by jeremyg-lunarg** [comment]
Sascha, FYI there will be several PRs after https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/2554 .  All of the changes we plan on merging are in the jeremyg-synchronization2 branch on github, but expect force rebases  on that branch...

**Comment by marty-johnson59** [comment]
Hi Sascha, I think we're getting close on this one.  I've added a bunch of reviewers. Can you resolve the conflicts here so we can merge once we get the approvals?  Thx

**Comment by SaschaWillems** [comment]
Sure, just merged with master from upstream to resolve the conflicts.

**Comment by marty-johnson59** [comment]
@AlejandroCosin, @aurolou, @carlosadcq, @gary-sweett, @gpx1000, @JoseEmilio-ARM, @r-potter, @sfricke-samsung, @wasimabbas-arm, @ZandroFargnoli

Hi reviewers, we're about to go live with the SDK and blog for Sync2 (targeting Monday).  Would be cool if we could include Sascha's sample at the same time.  Any chance you can break free to check this out?  Thanks!

mcj

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:138]: I would suggest to prefer C++ style casting over C style.  Just a small nitpick.
`VkViewport viewport = vkb::initializers::viewport(static_cast<float>(width), static_cast<float>(height), 0.0f, 1.0f);`

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:271]: FWIW, this loop has what looks like several transitory calculations that could be brought out of the loops as an optimization.  It's probably better to leave as is for readability.

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:25]: constructor doesn't initialize num_particles

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:143]: More nitpicking: NULL -> nullptr for C++

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:203]: The final 0 should be a nullptr; while it's effectively the same, in a sample it's best to show best practices and be precise.

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:231]: declaration shadows a previous variable on line 203

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:241]: dependency_info shadows a previous definition on line 217

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:406]: NULL -> nullptr

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:459]: uninitialized variable.

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:597]: uninitialized (put a {} on 597)
`} specialization_data{};`

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:581]: NULL -> nullptr

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:666]: command_buffer_allocate_info shadows a previous variable name on line 637.

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.cpp:709]: variable name shadows a previously defined one on line 650

File: samples/extensions/synchronization_2/synchronization_2.h
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.h:39]: I personally prefer to use initializer lists for easier reference (everything is in the cpp file for the construction; but I realize this is a personal preference; num_particles isn't inited).

File: samples/extensions/synchronization_2/synchronization_2.h
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.h:99]: virtual destructor.  mark with override.

File: samples/extensions/synchronization_2/synchronization_2.h
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.h:100]: remove virtual keep override

File: samples/extensions/synchronization_2/synchronization_2.h
Comment (gpx1000) [inline samples/extensions/synchronization_2/synchronization_2.h:116]: remove virtual keep override

File: shaders/synchronization_2/particle.vert
Comment (gpx1000) [inline shaders/synchronization_2/particle.vert:42]: add a newline at end of file

File: shaders/synchronization_2/particle_calculate.comp
Comment (gpx1000) [inline shaders/synchronization_2/particle_calculate.comp:89]: add newline at end of file

File: shaders/synchronization_2/particle_integrate.comp
Comment (gpx1000) [inline shaders/synchronization_2/particle_integrate.comp:48]: add newline

**Review by gpx1000 — DISMISSED** [review]
Everything looks good.  it worked well for me in Linux.
I added some minor requests in the code.  None of them are worth holding back an approval.  Fixing them are optional.

**Comment by SaschaWillems** [comment]
Thanks a lot for the exhaustive code comments @gpx1000. They all do make sense, and also apply to the compute N-Body sample this is based on. Once this sample has been merged, I'll do another PR for your suggestions.

**Review by carlosadcq — DISMISSED** [review]

File: samples/extensions/synchronization_2/synchronization_2.h
Comment (mklefrancois) [inline samples/extensions/synchronization_2/synchronization_2.h:117]: Remove redundant 'virtual'

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (mklefrancois) [reply samples/extensions/synchronization_2/synchronization_2.cpp:459]: I think this is fine for readability. The next two lines are initializing `shader_stages`

File: samples/extensions/synchronization_2/synchronization_2.cpp
Comment (mklefrancois) [reply samples/extensions/synchronization_2/synchronization_2.cpp:597]: And to initialize the struct with the same values as in `particle_calulate.comp`
```
        struct SpecializationData
	{
		uint32_t workgroup_size{0};
		uint32_t shared_data_size{1024};
		float    gravity{0.002f};
		float    power{0.75f};
		float    soften{0.05f};
	} specialization_data;
```

**Review by mklefrancois — DISMISSED** [review]
Work and tested with
- NVIDIA A6000
- Windows Driver 457.96
- Vulkan 1.2.170.0

**Pull request merged**
